homeassistant/components/iotty/coordinator.py

```diff
@@ -0,0 +1,90 @@
+"""DataUpdateCoordinator for iotty."""
+
+from dataclasses import dataclass
+import logging
+
+from iottycloud.device import Device
+from iottycloud.verbs import RESULT, STATUS
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.config_entry_oauth2_flow import OAuth2Session
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from . import api
+from .const import DOMAIN, LOGGER, PLATFORMS, UPDATE_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class IottyData:
+    """iotty data stored in the DataUpdateCoordinator."""
+
+    devices: list[Device]
+
+
+class IottyDataUpdateCoordinator(DataUpdateCoordinator[IottyData]):
+    """Class to manage fetching Iotty data."""
+
+    config_entry: ConfigEntry
+    _entities: dict
+    _devices: list[Device]
+
+    def __init__(
+        self, hass: HomeAssistant, entry: ConfigEntry, session: OAuth2Session
+    ) -> None:
+        """Initialize the coordinator."""
+        _LOGGER.debug("Initializing iotty data update coordinator")
+
+        super().__init__(
+            hass,
+            LOGGER,
+            name=f"{DOMAIN}_coordinator",
+            update_interval=UPDATE_INTERVAL,
+        )
+
+        self.config_entry = entry
+        self._entities = {}
+        self._devices = []
+        self.iotty = api.IottyProxy(
+            hass, aiohttp_client.async_get_clientsession(hass), session
+        )
+
+    def store_entity(self, device_id: str, entity: Entity) -> None:
+        """Store iotty device within Hass entities."""
+        _LOGGER.debug("Storing device '%s' in entities", device_id)
+        self._entities[device_id] = entity
+
+    async def async_config_entry_first_refresh(self) -> None:
+        """Override the first refresh to also fetch iotty devices list."""
+        _LOGGER.debug("Fetching devices list from iottyCloud")
+        self._devices = await self.iotty.get_devices()
```

Comment:
Can this happen after the super() call? Also, why do we only get the device list once, don't we need to do that at every update and handle new devices?

Reply:
You need to fix it such that new devices are automatically added, here's an example of how it can be implemented: https://github.com/home-assistant/core/blob/e472af6fb453784702dc5510dd02ce5c7257a231/homeassistant/components/meater/sensor.py#L146-L174

About removed devices, it depends on if the user needs to manually remove already installed devices from the iotty supporting app or not. If they do, and already installed devices are always included in the API responses regardless of if they lose power or can't connect to the WiFi network you should automatically clean up such devices by adding another listener to the coordinator which removed devices removed by the user.

If devices are not included in API responses when they lose power, you should instead allow the user to remove devices by implementing a function like this in `__init__.py`
```py
async def async_remove_config_entry_device(
    hass: HomeAssistant, config_entry: ConfigEntry, device_entry: DeviceEntry
) -> bool:
```

Reply:
Hi, I have followed your suggestions, let me know if that is ok and satisfies your request

Reply:
It looks good, except that you're not handling removed devices.

Reply:
Hi, in my case the API will not return any removed device, so according to your previous comment (https://github.com/home-assistant/core/pull/103073#discussion_r1613504099), I should be in this situation:

> If devices are not included in API responses when they lose power, you should instead allow the user to remove devices by implementing a function like this in __init__.py

This is what I did. The function is empty because I do not need to perform further action, since the API is already behaving as the device does not exist, so the only thing left is removing the device from the entity.
According to the official docs (https://developers.home-assistant.io/docs/device_registry_index?_highlight=async_remove_config_entry_device#removing-devices) it should be done automatically by hass.

> When the user clicks the delete device button for the device and confirms it, async_remove_config_entry_device will be awaited and if True is returned, the config entry will be removed from the device. If it was the only config entry of the device, the device will be removed from the device registry.

Looks to me like I did all of the things requested by you and the documentation, am I missing something? What kind of other cleanup am I supposed to do?

Reply:
I'm confused, you write:

> Hi, in my case the API will not return any removed device, so according to your previous comment (https://github.com/home-assistant/core/pull/103073#discussion_r1613504099), I should be in this situation:

>> If devices are not included in API responses when they lose power, you should instead allow the user to remove devices by implementing a function like this in __init__.py

Let me try to rephrase:

If a device, which has previously been included in API responses, is disconnected from power but the user takes no action to remove it via the iotty app, will that device be included in or omitted from API-responses?

Reply:
This was discussed and resolved in a Discord chat